packages/versioned-components/src/predicates.ts

```diff
@@ -0,0 +1,99 @@
+// We need several kinds of guards:
+// 1) `guard.plain(version)`, to check the version string
+// 2) `guard.props({ version: ... })`, to check and narrow the props or packages object
+// 3) `guard.propsByVersion(packages, props)`, to incrementally update the props
+// The following code is a bit complicated, because it optimizes for easily adding new guards.
+
+import { type FC } from "react";
+
+import { type SquigglePackages } from "./versionedSquigglePackages.js";
+import { type SquiggleVersion } from "./versions.js";
+
+type VersionGuard<AllowedSquiggleVersion extends SquiggleVersion> = (
+  version: SquiggleVersion
+) => version is AllowedSquiggleVersion;
+
+type VersionedObjectGuard<AllowedSquiggleVersion extends SquiggleVersion> = <
+  Obj extends { version: SquiggleVersion },
+>(
+  obj: Obj
+) => obj is Extract<Obj, { version: AllowedSquiggleVersion }>;
+
+type ExtractKeys<T extends object, V> = {
+  [K in keyof T]: T[K] extends V ? K : never;
+}[keyof T];
+
+// TODO - include components for all versions
```

Comment:
This might be a problem in the future. I think this type does an _intersection_ of component names from all versions, instead of a union. So `versionSupportsFoo.propsByVersion<"MyNewSquiggleComponent">(...)` won't be available.

I'll figure it out when it becomes a problem.